src/helpers/data_normalization.ts

```diff
@@ -1,25 +1,32 @@
 import { Position } from "../types";
-import { deepEquals, largeMax } from "./misc";
 import { recomputeZones } from "./recompute_zones";
 import { positionToZone, zoneToXc } from "./zones";
 
+type ReverseLookup = Map<string, number>;
+type ItemsDic<T> = { [id: number]: T };
+const globalReverseLookup = new Map<ItemsDic<any>, ReverseLookup>();
+
 /**
  * Get the id of the given item (its key in the given dictionary).
  * If the given item does not exist in the dictionary, it creates one with a new id.
  */
-export function getItemId<T>(item: T, itemsDic: { [id: number]: T }) {
-  for (const key in itemsDic) {
-    if (deepEquals(itemsDic[key], item)) {
-      return parseInt(key, 10);
-    }
+export function getItemId<T>(item: T, itemsDic: ItemsDic<T>) {
+  if (!globalReverseLookup.has(itemsDic)) {
+    globalReverseLookup.set(itemsDic, new Map());
```

Comment:
Actually, it's not ok like that.
It's a pure memory leak because a new cache is added for every new `itemsDic` (=everytime we export). It's specially problematic in tests where we export a lot of times.

Could you try a class alternative ?